File: 0815-bus-routes/0815-bus-routes.py
class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        if source == target:
            return 0
        stop_to_buses = defaultdict(list)


        for i, stops in enumerate(routes):
            for stop in stops:
                stop_to_buses[stop].append(i)

        vis = set()
        q = deque()
        for bus in stop_to_buses[source]:
            q.append(bus)
            vis.add(bus)

        num_buses = 1

        while q:
            sz = len(q)

            for _ in range(sz):
                bus_no = q.popleft()

                for stop in routes[bus_no]:
                    if stop == target:
                        return num_buses

                    for bus in stop_to_buses[stop]:
                        if bus not in vis:
                            q.append(bus)
                            vis.add(bus)

            num_buses += 1

        return -1
                

File: 0815-bus-routes/0815-bus-routes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        if (source == target) return 0;

        // 1. Map each stop to the buses that visit it
        Map<Integer, List<Integer>> stopsToBuses = new HashMap();
        for (int bus = 0; bus < routes.length; bus++) {
            for (int stop : routes[bus]) {
                stopsToBuses.computeIfAbsent(stop, k -> new ArrayList()).add(bus);
            }
        }

        // 2. Initialize BFS
        Queue<int[]> q = new LinkedList<>();
        Set<Integer> visitedBuses = new HashSet<>();
        
        if (!stopsToBuses.containsKey(source)) return -1;
        
        for (int busIdx : stopsToBuses.get(source)) {
            q.add(new int[]{busIdx, 1}); // {bus_index, count_of_buses_taken}
            visitedBuses.add(busIdx);
        }

        // 3. Perform BFS
        while (!q.isEmpty()) {
            int[] hopInfo = q.poll();
            int currentBus = hopInfo[0];
            int busCount = hopInfo[1];

            for (int stop : routes[currentBus]) {
                if (stop == target) return busCount;

                // Add all unvisited buses passing through this stop
                for (int nextBus : stopsToBuses.get(stop)) {
                    if (!visitedBuses.contains(nextBus)) {
                        visitedBuses.add(nextBus);
                        q.add(new int[]{nextBus, busCount + 1});
                    }
                }
            }
        }

        return -1;
    }
}